File: Leetcode Challenge/February/cpp/Merge k Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include<queue>
class Solution {
public:
    
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<int, vector<int>, greater<int>> pq; //min heap
        
        if(lists.size()==0){
            return NULL;
        }
        if(lists[0]==NULL && lists.size()==1){
            return NULL;
        }
        for(int i=0;i<lists.size();i++){
            ListNode *head=lists[i];
            if(head==NULL){
                continue;
            }
            ListNode *temp=head;
            while(temp!=NULL){
                pq.push(temp->val);
                temp=temp->next;
            }
        }
        
        if(pq.size()==0){
            return NULL;
        }
    
        ListNode *newNode=new ListNode(pq.top());
        pq.pop();
        ListNode *finalHead=newNode;
        ListNode *tail=finalHead;
        while(!pq.empty()){
            ListNode *nextNode=new ListNode(pq.top());
            pq.pop();
            tail->next=nextNode;
            tail=tail->next;
            
        }
        
        tail->next=NULL;
        return finalHead;
        
    }
};

File: Leetcode Challenge/February/java/Merge k Sorted Lists.java
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Queue<ListNode> q = new PriorityQueue<ListNode>((a,b) -> a.val - b.val);
        // only head pointer of all lists are added
        for(ListNode l : lists){
            if(l!=null){
                q.add(l);
            }        
        }
        ListNode head = new ListNode(-1);
        ListNode point = head;
        while(!q.isEmpty()){ 
            point.next = q.poll();
            point = point.next; 
            if(point.next!=null){
                q.add(point.next); // add next of current smallest
            }
        }
        return head.next;
    }
}
